tool/planet/agent.go

```diff
@@ -128,151 +122,86 @@ func startLeaderClient(conf *LeaderConfig, agent agent.Agent, errorC chan error)
 
-		if newVal != conf.PublicIP {
-			return
-		}
+	return &clientState{
+		client:     client,
+		reconciler: reconciler,
+		mon:        mon,
+	}, nil
+}
 
-		ctx, cancel := context.WithTimeout(context.Background(), recordEventTimeout)
-		defer cancel()
+// Close closes this state by releasing the resources.
+// Implements io.Closer
+func (r *clientState) Close() error {
+	r.reconciler.Stop()
+	r.mon.close()
+	return r.client.Close()
+}
 
-		agent.RecordLocalEvents(ctx, []*pb.TimelineEvent{
-			pb.NewLeaderElected(agent.GetConfig().Clock.Now(), prevVal, newVal),
-		})
-	})
+type clientState struct {
+	client     *leader.Client
+	reconciler *reconcile.Reconciler
+	mon        *unitMonitor
+}
 
-	var cancelVoter context.CancelFunc
-	var ctx context.Context
-	if conf.Role == RoleMaster {
-		switch conf.ElectionEnabled {
-		case true:
-			log.Infof("adding voter for IP %v", conf.PublicIP)
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
-				return nil, trace.Wrap(err)
-			}
-		case false:
-			log.Info("shutting down services until election has been re-enabled")
-			// Shut down services at startup if running as master
-			if err := unitsCommand("stop"); err != nil {
-				log.Infof("failed to stop units: %v", err)
-			}
-		}
-	}
-	// watch the election mode status and start/stop participation
-	// depending on the value of the election mode key
-	client.AddWatchCallback(conf.ElectionKey, conf.Term, func(key, prevVal, newVal string) {
-		var err error
-		enabled, _ := strconv.ParseBool(newVal)
-		switch enabled {
-		case true:
-			if cancelVoter != nil {
-				log.Infof("voter is already active")
-				return
-			}
-			// start election participation
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
-				log.Errorf("failed to add voter for %v: %v", conf.PublicIP, trace.DebugReport(err))
-				errorC <- err
-			}
-		case false:
-			if cancelVoter == nil {
-				log.Info("no voter active")
-				return
-			}
-			// stop election participation
-			cancelVoter()
-			cancelVoter = nil
+func updateLeaderTrigger(ctx context.Context, agent agent.Agent, reconciler *reconcile.Reconciler, state *agentState, logger log.FieldLogger) leader.CallbackFn {
+	return func(_, prevValue, newLeaderAddr string) {
+		if newLeaderAddr == "" || prevValue == newLeaderAddr {
+			return
 		}
-	})
-	// modify /etc/hosts upon election of a new leader node
-	client.AddWatchCallback(conf.LeaderKey, conf.Term/3, func(key, prevVal, newLeaderIP string) {
-		if err := updateDNS(conf, hostname, newLeaderIP); err != nil {
-			log.Error(trace.DebugReport(err))
+		logger.WithField("leader-addr", newLeaderAddr).Info("New leader.")
+		plan := state.withLeaderAddr(newLeaderAddr)
+		reconciler.Reset(ctx, plan)
+		if newLeaderAddr != plan.localAddr {
```

Comment:
I think we want `if newLeaderAddr == plan.localAddr` here. We only want the new leader that was elected to record a new event.